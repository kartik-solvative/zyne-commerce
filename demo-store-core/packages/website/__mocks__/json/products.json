[
  {
    "productCode": "ABC",
    "variantCode": "ABCD",
    "sku": "ABCD-INVALID-VARIANT-TYPE",
    "slug": "/black-five-panel-cap-with-white-logo/5PANECAP000000FFFFFFXXXX",
    "name": {
      "en": "Black Five-Panel Cap with White Logo",
      "it": "Black Five-Panel Cap with White Logo"
    },
    "description": {
      "en": "Soft-structured, five-panel, low-profile cap. 100% cotton, metal eyelets, nylon strap clip closure.",
      "it": "Soft-structured, five-panel, low-profile cap. 100% cotton, metal eyelets, nylon strap clip closure."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/5PANECAP000000FFFFFFXXXX_FLAT.png"
    ],
    "color": ["000000"]
  },
  {
    "productCode": "BODYBSSS",
    "variantCode": "BODYBSSS000000",
    "sku": "BODYBSSS000000FFFFFF12MX",
    "slug": "/black-baby-short-sleeve-bodysuit-with-white-logo-12-months/BODYBSSS000000FFFFFF12MX",
    "facetA": "000000",
    "color": "000000",
    "size": "12 months",
    "name": {
      "en-US": "Translation for en-US - Black Baby Short Sleeve Bodysuit with White Logo",
      "en": "Translation for en - Black Baby Short Sleeve Bodysuit with White Logo",
      "it": "Translation for it - Body nero a maniche corte per bebè con logo bianco"
    },
    "description": {
      "en-US": "Translation for en-US - This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.",
      "en": "Translation for en - This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.",
      "it": "Translation for it - This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/BODYBSSS000000FFFFFF12MX_FLAT.png"
    ]
  },
  {
    "productCode": "BODYBSSS",
    "variantCode": "BODYBSSS000000",
    "sku": "BODYBSSS000000FFFFFF6MXX",
    "slug": "/black-baby-short-sleeve-bodysuit-with-white-logo-6-months/BODYBSSS000000FFFFFF6MXX",
    "facetA": [],
    "color": "000000",
    "size": "6 months",
    "name": {
      "en": "Black Baby Short Sleeve Bodysuit with White Logo",
      "it": "Body nero a maniche corte per bebè con logo bianco"
    },
    "description": {
      "en": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.",
      "it": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/BODYBSSS000000FFFFFF6MXX_FLAT.png"
    ]
  },
  {
    "productCode": "BODYBSSS",
    "variantCode": "BODYBSSSFFFFFF",
    "sku": "BODYBSSSFFFFFF00000012MX",
    "slug": "/white-baby-short-sleeve-bodysuit-with-black-logo-12-months/BODYBSSSFFFFFF00000012MX",
    "facetA": ["FFFFFF"],
    "color": "FFFFFF",
    "size": "12 months",
    "name": {
      "en": "White Baby Short Sleeve Bodysuit with Black Logo",
      "it": "Body bianco a maniche corte per bebè con logo nero"
    },
    "description": {
      "en": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.",
      "it": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/BODYBSSSFFFFFF00000012MX_FLAT.png"
    ]
  },
  {
    "productCode": "BODYBSSS",
    "variantCode": "BODYBSSSFFFFFF",
    "sku": "BODYBSSSFFFFFF0000006MXX",
    "slug": "/white-baby-short-sleeve-bodysuit-with-black-logo-6-months/BODYBSSSFFFFFF0000006MXX",
    "facetA": ["FFFFFF"],
    "color": "FFFFFF",
    "size": "6 months",
    "name": {
      "en": "White Baby Short Sleeve Bodysuit with Black Logo",
      "it": "Body bianco a maniche corte per bebè con logo nero"
    },
    "description": {
      "en": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.",
      "it": "This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/BODYBSSSFFFFFF0000006MXX_FLAT.png"
    ]
  },
  {
    "productCode": "BOTT17OZ",
    "variantCode": "BOTT17OZFFFFFF",
    "sku": "BOTT17OZFFFFFF000000XXXX",
    "slug": "/white-water-bottle-with-black-logo-17oz/BOTT17OZFFFFFF000000XXXX",
    "facetA": ["FFFFFF"],
    "color": "FFFFFF",
    "name": {
      "en": "White Water Bottle with Black Logo",
      "it": "Bottiglia d'acqua bianca con logo nero"
    },
    "description": {
      "en": "Beautifully designed water bottle, made from high-grade stainless steel. Its double-wall construction makes it perfect for storing both hot and cold liquids.",
      "it": "Beautifully designed water bottle, made from high-grade stainless steel. Its double-wall construction makes it perfect for storing both hot and cold liquids."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/BOTT17OZFFFFFF000000XXXX_FLAT.png"
    ]
  },
  {
    "productCode": "CUFFBEAN",
    "variantCode": "CUFFBEAN000000",
    "sku": "CUFFBEAN000000FFFFFFXXXX",
    "slug": "/black-cuffed-beanie-with-white-logo/CUFFBEAN000000FFFFFFXXXX",
    "facetA": ["000000"],
    "color": "000000",
    "name": {
      "en": "Black Cuffed Beanie with White Logo",
      "it": "Berretto con risvolto nero con logo bianco"
    },
    "description": {
      "en": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe.",
      "it": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/CUFFBEAN000000FFFFFFXXXX_FLAT.png"
    ]
  },
  {
    "productCode": "CUFFBEAN",
    "variantCode": "CUFFBEANA6A5A0",
    "sku": "CUFFBEANA6A5A0000000XXXX",
    "slug": "/gray-cuffed-beanie-with-black-logo/CUFFBEANA6A5A0000000XXXX",
    "facetA": ["A6A5A0"],
    "color": "A6A5A0",
    "name": {
      "en": "Gray Cuffed Beanie with Black Logo",
      "it": "Berretto con risvolto grigio con logo nero"
    },
    "description": {
      "en": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe.",
      "it": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/CUFFBEANA6A5A0000000XXXX_FLAT.png"
    ]
  },
  {
    "productCode": "CUFFBEAN",
    "variantCode": "CUFFBEANFFFFFF",
    "sku": "CUFFBEANFFFFFF000000XXXX",
    "slug": "/white-cuffed-beanie-with-black-logo/CUFFBEANFFFFFF000000XXXX",
    "facetA": ["FFFFFF"],
    "color": "FFFFFF",
    "name": {
      "en": "White Cuffed Beanie with Black Logo",
      "it": "Berretto con risvolto bianco con logo nero"
    },
    "description": {
      "en": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe.",
      "it": "A snug, form-fitting beanie. It's not only a great head-warming piece but a staple accessory in anyone's wardrobe."
    },
    "images": [
      "https://data.commercelayer.app/seed/images/skus/CUFFBEANFFFFFF000000XXXX_FLAT.png"
    ]
  }
]
