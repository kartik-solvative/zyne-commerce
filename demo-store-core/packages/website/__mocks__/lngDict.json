{
  "general": {
    "viewAll": "View all",
    "search": "search",
    "addToCart": "Add to cart",
    "chooseCountry": "Choose country",
    "international": "International",
    "resultsFor": "{{count}} results for \"{{query}}\""
  },
  "seo": {
    "title": "Commerce Layer Store",
    "description": "Commerce Layer is a transactional commerce API that lets you add multi-market ecommerce to any digital experience, with ease."
  },
  "search": {
    "variants": {
    "color": "Colors",
    "size": "Choose size"
    },
    "facets": {
      "gender": "Gender",
      "category": "Category",
      "price": {
        "amount_float": "Price"
      },
      "available": "Show only available",
      "color": "Color",
      "size": "Size"
    },
    "values": {
      "#000000": "Black",
      "#FFFFFF": "White",
      "#C2C2C2": "Grey",
      "#E7DDC7": "Beige"
    }
  },
  "product": {
    "description": "Product info"
  }
}
