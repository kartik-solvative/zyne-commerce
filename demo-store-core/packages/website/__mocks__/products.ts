import { rawDataProducts_schema } from '@commercelayer/demo-store-types'
import productsJson from '#__mocks__/json/products.json'

export default rawDataProducts_schema.parse(productsJson)