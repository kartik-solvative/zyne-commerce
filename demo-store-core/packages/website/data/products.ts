import { fetchJsonData } from '#utils/data'
import { memoize } from '#utils/memoize'
import { RawDataProduct as BaseRawDataProduct, rawDataProducts_schema } from '@commercelayer/demo-store-types'
import type { Price } from '@commercelayer/sdk'

export const getRawDataProducts = memoize(
  async function(): Promise<RawDataProduct[]> {
    // TODO: this should be unserializable
    return rawDataProducts_schema.parse(
      await fetchJsonData('products')
    )
  }
)

export type RawDataProduct = BaseRawDataProduct & {
  available?: boolean
  price?: Price
}
