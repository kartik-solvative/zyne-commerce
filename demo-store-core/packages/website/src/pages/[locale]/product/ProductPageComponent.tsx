import { Accordion } from '#components/Accordion'
import { Carousel } from '#components/Carousel'
import { Price } from '#components/CommerceLayer/Price'
import type { HeaderProps } from '#components/Header'
import { Page } from '#components/Page'
import { VariantSelector } from '#components/VariantSelector'
import { useAuthContext } from '#contexts/AuthContext'
import { useSettingsContext } from '#contexts/SettingsContext'
import type { LocalizedProduct, LocalizedProductWithVariant } from '#utils/products'
import { getProductUrl } from '#utils/url'
import { AddToCartButton, AvailabilityContainer, AvailabilityTemplate } from '@commercelayer/react-components'
import type { NextPage } from 'next'
import { useI18n } from 'next-localization'
import { useMemo, useState } from 'react'

export type Props = HeaderProps & {
  product: LocalizedProductWithVariant
  localeCodes: string[]
}

export const ProductPageComponent: NextPage<Props> = ({ navigation: links, product, localeCodes }) => {
  const i18n = useI18n()
  const [currentProduct, setCurrentProduct] = useState<LocalizedProduct>()

  const auth = useAuthContext()
  const settings = useSettingsContext()

  const slides = useMemo(() => product.images.map(image => <img key={image} src={image} alt={product.name} className='w-full' />), [product])

  return (
    <Page
      localeCodes={localeCodes}
      navigation={links}
      canonical={getProductUrl(product.variants[0])}
      title={product.name}
      description={product.description}
    >
      <div className='flex flex-col lg:flex-row gap-6 lg:gap-24 xl:gap-48 mt-12 lg:mt-24'>

        <div className=' basis-1/2'>
          <Carousel slides={slides} />
        </div>

        <div className='grow-0 shrink-1 basis-1/2'>
          <h1 className='text-3xl mb-6'>{product.name}</h1>

          <div className='min-h-[30px]'>
            <Price code={currentProduct?.sku} className='text-xl' />
          </div>

          <hr className='text-gray-100 my-8' />

          <VariantSelector product={product} onChange={setCurrentProduct} />

          {
            settings.locale?.isShoppable && auth.accessToken && (
              <>
                <AddToCartButton
                  skuCode={currentProduct?.sku}
                  label={i18n.t('general.addToCart')}
                  className='block w-full mt-12 h-14 px-6 font-semibold rounded-md text-white bg-red-800 disabled:bg-gray-300' />

                <AvailabilityContainer skuCode={currentProduct?.sku}>
                  <AvailabilityTemplate className='mt-6 hidden' showShippingMethodName={true} showShippingMethodPrice={true} color={'blue'} timeFormat={'days'} />
                </AvailabilityContainer>
              </>
            )
          }

          <div className='border-b border-b-gray-400 mt-12'>
            <Accordion data-testid='product-detail' title={<div className='font-extrabold'>{i18n.t('product.description')}</div>}>
              <div className='mt-6 mb-8'>{product.description}</div>
            </Accordion>

            {
              product.details?.map(detail => (
                <Accordion key={detail.title} data-testid='product-detail' title={<div className='font-extrabold'>{detail.title}</div>}>
                  <div className='mt-6 mb-8'>{detail.content}</div>
                </Accordion>
              ))
            }
          </div>
        </div>

      </div>
    </Page>
  )
}
