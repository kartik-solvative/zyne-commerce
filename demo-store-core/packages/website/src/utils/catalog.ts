import type { Catalog, Taxon, Taxonomy } from '#data/models/catalog'
import { deepFind, DeepFindResult } from '#utils/collection'
import { getSearchUrl } from '#utils/url'
import uniq from 'lodash/uniq'
import type { Unserializable } from './unserializable'

export type NavigationLink = {
  key: string
  href: string
  text: string
  description: string
  children?: NavigationLink[]
}

export type Breadcrumbs = {
  parent: NavigationLink[]
  current: NavigationLink
  children: NavigationLink[]
}

export type Navigation = {
  current: NavigationLink
  path: NavigationLink[]
}


const getPrimaryTaxonomy = (catalog: Unserializable<Catalog>): Taxonomy => {
  // TODO: taxonomies[0] is a requirement. First taxonomy is considered the navigation one. Should it be configurable?
  return catalog.value.taxonomies[0]
}

export const getRootNavigationLinks = (catalog: Unserializable<Catalog>): NavigationLink[] => {
  return getPrimaryTaxonomy(catalog).taxons.map(taxon => ({
    key: taxon.id,
    href: getSearchUrl(taxon),
    text: taxon.label,
    description: taxon.description
  }))
}

export const getBreadcrumbs = (taxon: DeepFindResult<Taxon>): Breadcrumbs => {
  return {
    parent: taxon.memo.map(({ id: key, slug, label, description }) => ({
      key: key,
      href: getSearchUrl({slug}),
      text: label,
      description
    })),
    current: {
      key: taxon.result.id,
      href: getSearchUrl(taxon.result),
      text: taxon.result.label,
      description: taxon.result.description
    },
    children: taxon.result.taxons.map(({ id: key, slug, label, description }) => ({
      key: key,
      href: getSearchUrl({slug}),
      text: label,
      description
    }))
  }
}

const getNavigationChildren = (foundTaxon: DeepFindResult<Taxon>, index: number, currentTaxon: Taxon): NavigationLink[] => {
  return foundTaxon.memo[index]?.id === currentTaxon.id ? currentTaxon.taxons.filter(t => flattenReferencesFromTaxon(t).length > 0).map((taxon) => ({
    key: taxon.id,
    href: getSearchUrl({ slug: taxon.slug }),
    text: taxon.label,
    description: taxon.description,
    children: getNavigationChildren(foundTaxon, ++index, taxon)
  })) : []
}

export const getNavigation = (foundTaxon: DeepFindResult<Taxon>): Navigation => {
  return {
    current: {
      key: foundTaxon.result.id,
      href: getSearchUrl(foundTaxon.result),
      text: foundTaxon.result.label,
      description: foundTaxon.result.description
    },
    path: [{
      key: foundTaxon.memo[0].id,
      href: getSearchUrl(foundTaxon.memo[0]),
      text: foundTaxon.memo[0].label,
      description: foundTaxon.memo[0].description,
      children: getNavigationChildren(foundTaxon, 0, foundTaxon.memo[0])
    }]
  }
}

export const getSlugs = (catalog: Unserializable<Catalog>): string[] => {
  function getFlatSlug(taxon: Taxon): string[] {
    return [taxon.slug].concat(taxon.taxons.flatMap(getFlatSlug))
  }

  return getPrimaryTaxonomy(catalog).taxons.flatMap(getFlatSlug)
}

export function flattenReferencesFromTaxon(taxon: Taxon): string[] {
  return uniq(
    taxon.references.concat(taxon.taxons?.flatMap(flattenReferencesFromTaxon) || [])
  )
}

export function flattenReferencesFromCatalog(catalog: Unserializable<Catalog>): string[] {
  return uniq(
    catalog.value.taxonomies.flatMap(({ taxons }) => taxons.flatMap(flattenReferencesFromTaxon)),
  )
}

export function findTaxonBySlug(catalog: Unserializable<Catalog>, slug: string): DeepFindResult<Taxon> {
  const taxon = catalog.value.taxonomies.reduce((acc, cv) => {
    if (acc) {
      return acc
    }

    return deepFind(cv.taxons, 'taxons', 'slug', slug)
  }, undefined as DeepFindResult<Taxon> | undefined)

  if (!taxon) {
    throw new Error('Cannot find Taxon!')
  }

  return taxon
}
