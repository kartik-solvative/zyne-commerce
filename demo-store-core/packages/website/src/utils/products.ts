import variantsConfig from '#config/variants.config'
import { translateField } from '#utils/locale'
import type { RawDataProduct } from '#data/products'
import get from 'lodash/get'
import uniqBy from 'lodash/uniqBy'

export type Variant = {
  name: string
  value: string
}

export type LocalizedProduct = Omit<RawDataProduct, 'name' | 'description' | 'variant' | 'details'> & {
  _locale: string
  name: string
  description: string
  variant: Variant[]

  // TODO: this should be removed when PLP
  details: {
    title: string
    content: string
  }[]
}

export type LocalizedProductWithVariant = LocalizedProduct & {
  variants: LocalizedProduct[]
}



function resolveProductLocale(product: RawDataProduct | LocalizedProduct | LocalizedProductWithVariant, locale: string): LocalizedProduct | LocalizedProductWithVariant {
  if ('_locale' in product) {
    return {
      ...product
    }
  }

  const variant: Variant[] = []

  variantsConfig.forEach(config => {
    if (config.field in product) {
      const value = get(product, config.field)

      if (typeof value !== 'string') {
        throw new Error(`The variant property "${config.field}" for the product ${product.sku} must be a string. Found ${JSON.stringify(value)}.`)
      }

      variant.push({
        name: config.field,
        value
      })
    }
  })

  return {
    ...product,
    _locale: locale,
    name: translateField(product, 'name', locale),
    description: translateField(product, 'description', locale),
    variant,
    details: product.details?.map(detail => ({
      title: translateField(detail, 'title', locale),
      content: translateField(detail, 'content', locale)
    })) || []
  }
}

export function getProduct(sku: string, locale: string, productList: (LocalizedProduct | RawDataProduct)[]): LocalizedProduct {
  const product = productList.find(product => product.sku === sku)

  if (!product) {
    throw new Error(`Cannot find a Product with sku equal to ${sku}`)
  }

  return resolveProductLocale(product, locale)
}

function getProductVariants(product: LocalizedProduct, productList: (RawDataProduct | LocalizedProduct | LocalizedProductWithVariant)[]): LocalizedProduct[] {
  return productList
    .filter(p => p.productCode === product.productCode)
    .map(p => {
      const localizedProduct = resolveProductLocale(p, product._locale)

      if ('variants' in localizedProduct) {
        // @ts-expect-error
        delete localizedProduct['variants']
        return localizedProduct as LocalizedProduct
      }

      return localizedProduct
    })
}

export function getProductWithVariants(sku: string, locale: string, productList: (LocalizedProduct | RawDataProduct)[]): LocalizedProductWithVariant {
  const product = getProduct(sku, locale, productList)
  const variants = getProductVariants(product, productList)

  return {
    ...product,
    variants
  }
}

export function addProductVariants(product: LocalizedProduct, productList: (LocalizedProduct | RawDataProduct)[]): LocalizedProductWithVariant {
  const variants = getProductVariants(product, productList)

  return {
    ...product,
    variants
  }
}

/**
 * Given a list of products, it moves all the variants at top level.
 */
export function spreadProductVariants(products: LocalizedProductWithVariant[]): LocalizedProductWithVariant[] {
  return uniqBy(
    products.reduce((flattenProductWithVariants, product) => {
      flattenProductWithVariants.push(
        ...product.variants.map(v => ({ ...v, variants: product.variants }))
      )
      return flattenProductWithVariants
    }, [] as LocalizedProductWithVariant[]),
    'sku'
  )
}
