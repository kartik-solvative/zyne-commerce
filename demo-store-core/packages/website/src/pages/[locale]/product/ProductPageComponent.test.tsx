import { getRootNavigationLinks } from '#utils/catalog'
import { getProductWithVariants } from '#utils/products'
import lngDict from '#__mocks__/lngDict.json'
import productsJson from '#__mocks__/products'
import { act, render, screen } from '@testing-library/react'
import { createCatalog, createRouter, createTaxon, createTaxonomy } from 'jest.helpers'
import { I18nProvider } from 'next-localization'
import { ProductPageComponent } from './ProductPageComponent'

jest.mock('#data/languages', () => ({
  getRawDataLanguages: () => ([
    {
      "name": "English",
      "code": "en",
      "catalog": "AMER"
    },
    {
      "name": "Italiano",
      "code": "it",
      "catalog": "AMER"
    }
  ])
}))

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

beforeEach(() => {
  jest.useFakeTimers();

  useRouter.mockReset()
})

afterEach(() => {
  jest.clearAllTimers()
});

test('product detail page', async () => {
  const locale = 'en'
  const product = {
    ...getProductWithVariants('BODYBSSS000000FFFFFF12MX', locale, productsJson),
    details: [
      {
        title: 'Details',
        content: 'Text for "details"'
      }
    ]
  }


  useRouter.mockImplementation(() => createRouter(`/${product.slug}`))

  const taxon = createTaxon('1')
  const taxonomy = createTaxonomy('1', [taxon])
  const catalog = createCatalog('1', [taxonomy])
  const navigation = getRootNavigationLinks(catalog)

  let container
  await act(async () => {
    ({ container } = render(
      <I18nProvider lngDict={lngDict} locale={locale}>
        <ProductPageComponent localeCodes={['en', 'en-US', 'it-IT']} navigation={navigation} product={product} />
      </I18nProvider>
    ))

    jest.runAllTimers();
  })


  const productDetails = screen.queryAllByTestId('product-detail')

  expect(productDetails.length).toStrictEqual(2)

  expect(productDetails[0].querySelector('[data-testid="accordion-title"]')).toHaveTextContent('Product info+')
  expect(productDetails[1].querySelector('[data-testid="accordion-title"]')).toHaveTextContent('Details')

  expect(productDetails[0].querySelector('[data-testid="accordion-content"]')).toHaveTextContent(product.description)
  expect(productDetails[1].querySelector('[data-testid="accordion-content"]')).toHaveTextContent('Text for "details"')

  expect(container).toMatchSnapshot()
})
