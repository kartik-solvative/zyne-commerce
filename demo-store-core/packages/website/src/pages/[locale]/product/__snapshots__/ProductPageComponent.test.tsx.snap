// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product detail page 1`] = `
<div>
  <main
    class="flex flex-col h-screen"
  >
    <div
      class="container mx-auto px-6"
    >
      <header
        class="border-b-gray-200 border-b pb-6 sticky top-0 bg-pageBackground z-50"
      >
        <nav
          class="flex items-center justify-between flex-wrap mb-4"
        >
          <div
            class="flex items-center flex-no-shrink text-white mr-6"
          >
            <a
              href="/en-US"
            >
              <div
                class="text-black py-6 pb-6 "
              >
                <svg
                  aria-label="Commerce Layer logomark"
                  class="m-1 inline"
                  width="32"
                />
                <svg
                  aria-label="Commerce Layer logotype"
                  class="ml-2 hidden md:inline"
                  width="192"
                />
              </div>
            </a>
          </div>
          <div
            class="flex items-center w-auto flex-grow justify-end"
          />
        </nav>
        <div
          class="flex items-center justify-between flex-wrap gap-4 relative"
        >
          <div
            class="order-2 lg:order-1"
          >
            <nav
              class="hidden lg:block lg:relative absolute right-0 top-full w-full z-10 flex-col bg-pageBackground py-4 lg:py-0"
            >
              <a
                class="text-back font-bold py-2 px-3 my-2 mx-1 transition-colors duration-300 bg-transparent"
                href="/en-US/search/taxon-slug-1"
              >
                Label for 1
              </a>
            </nav>
            <button
              aria-controls="menu"
              aria-label="Open menu"
              class="w-11 h-11 flex flex-col items-center justify-center space-y-1 bg-black rounded block lg:hidden"
              type="button"
            >
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
            </button>
          </div>
          <div
            class="order-1 grow lg:grow-0"
          >
            <label
              class="relative py-2.5 rounded bg-gray-100 text-gray-400 focus-within:text-gray-600 block"
              for="email"
            >
              <svg
                class="pointer-events-none w-6 h-6 absolute top-1/2 transform -translate-y-1/2 left-3"
              />
              <form
                action="/en-US/search"
              >
                <input
                  class="form-input appearance-none bg-transparent w-full pl-14 focus:outline-none focus:shadow-outline"
                  placeholder="search"
                  value=""
                />
              </form>
            </label>
          </div>
        </div>
      </header>
      <div
        class="flex flex-col lg:flex-row gap-6 lg:gap-24 xl:gap-48 mt-12 lg:mt-24"
      >
        <div
          class=" basis-1/2"
        >
          <div
            class="swiper swiper-initialized swiper-horizontal carousel"
          >
            <div
              class="swiper-wrapper"
              style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
            >
              <div
                class="swiper-slide select-none"
              >
                <img
                  alt="Translation for en - Black Baby Short Sleeve Bodysuit with White Logo"
                  class="w-full"
                  src="https://data.commercelayer.app/seed/images/skus/BODYBSSS000000FFFFFF12MX_FLAT.png"
                />
              </div>
            </div>
            <div
              class="swiper-button-prev"
            />
            <div
              class="swiper-button-next"
            />
            <div
              class="swiper-pagination swiper-pagination-bullets swiper-pagination-horizontal"
            />
          </div>
        </div>
        <div
          class="grow-0 shrink-1 basis-1/2"
        >
          <h1
            class="text-3xl mb-6"
          >
            Translation for en - Black Baby Short Sleeve Bodysuit with White Logo
          </h1>
          <div
            class="min-h-[30px]"
          />
          <hr
            class="text-gray-100 my-8"
          />
          <div>
            <div
              class="my-6"
            >
              <p
                class="text-gray-800 text-lg font-semibold"
              >
                Colors
              </p>
              <button
                class="w-12 h-12 rounded-full inline-flex items-center justify-center my-1 mr-2 border border-gray-300 bg-gray-100"
              >
                <div
                  class="w-8 h-8 border border-gray-600 rounded-full bg-contain"
                >
                  <span
                    class="sr-only"
                  >
                    000000
                  </span>
                </div>
              </button>
              <button
                class="w-12 h-12 rounded-full inline-flex items-center justify-center my-1 mr-2 "
              >
                <div
                  class="w-8 h-8 border border-gray-600 rounded-full bg-contain"
                >
                  <span
                    class="sr-only"
                  >
                    FFFFFF
                  </span>
                </div>
              </button>
            </div>
            <div
              class="my-6"
            >
              <p
                class="text-gray-800 text-lg font-semibold"
              >
                Choose size
              </p>
              <button
                class="my-2 mr-2 rounded py-2 px-4 bg-black text-white"
              >
                12 months
              </button>
              <button
                class="my-2 mr-2 rounded py-2 px-4 bg-gray-100 text-gray-500"
              >
                6 months
              </button>
            </div>
          </div>
          <div
            class="border-b border-b-gray-400 mt-12"
          >
            <div
              class="flex flex-col"
              data-testid="product-detail"
            >
              <button
                class="accordion-title py-6 box-border border-t border-t-gray-400 appearance-none cursor-pointer focus:outline-none flex items-center justify-between"
                data-testid="accordion-title"
              >
                <div
                  class="inline-block text-footnote light"
                >
                  <div
                    class="font-extrabold"
                  >
                    Product info
                  </div>
                </div>
                <div
                  class="accordion-icon"
                >
                  +
                </div>
              </button>
              <div
                class="accordion-content overflow-hidden transition-all max-h-0 duration-400 ease-in-out"
                data-testid="accordion-content"
                style="max-height: 0px;"
              >
                <div
                  class="border-t border-t-gray-100"
                >
                  <div
                    class="mt-6 mb-8"
                  >
                    Translation for en - This comfortable baby one-piece is made of 100% combed ring-spun cotton except for heather grey color, which contains polyester. The lap shoulders ensure that the bodysuit can be easily put on and taken off, making for easy changing.
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
              data-testid="product-detail"
            >
              <button
                class="accordion-title py-6 box-border border-t border-t-gray-400 appearance-none cursor-pointer focus:outline-none flex items-center justify-between"
                data-testid="accordion-title"
              >
                <div
                  class="inline-block text-footnote light"
                >
                  <div
                    class="font-extrabold"
                  >
                    Details
                  </div>
                </div>
                <div
                  class="accordion-icon"
                >
                  +
                </div>
              </button>
              <div
                class="accordion-content overflow-hidden transition-all max-h-0 duration-400 ease-in-out"
                data-testid="accordion-content"
                style="max-height: 0px;"
              >
                <div
                  class="border-t border-t-gray-100"
                >
                  <div
                    class="mt-6 mb-8"
                  >
                    Text for "details"
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-grow py-8 lg:p-16 mt-24 bg-gray-50"
    >
      <div
        class="container mx-auto px-6"
      >
        <div
          class="lg:flex items-center justify-between flex-wrap"
        >
          <div>
            <div
              class="flex items-center justify-between"
            >
              <span
                class="flex items-center"
              >
                <svg
                  class="mr-1.5"
                />
                 
                International
              </span>
              <a
                class="uppercase text-violet-400 text-xs border-b border-gray-200 ml-3 mt-1"
                href="/"
              >
                Choose country
              </a>
            </div>
          </div>
          <div
            class="mt-6 lg:mt-0"
          >
            <select
              class=" inputSelect"
            >
              <option
                selected=""
                value="en"
              >
                English
              </option>
              <option
                value="it"
              >
                Italiano
              </option>
            </select>
          </div>
        </div>
        <div
          class="mt-6 py-6 border-y border-y-gray-100 text-xs flex flex-col md:flex-row md:divide-x"
        >
          <div
            class="py-2 md:py-0 md:pr-4"
          >
            <a
              href="/en-US/shipping-and-payments"
            >
              Shipping & Payments
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/returns"
            >
              Returns
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/terms-and-conditions"
            >
              Terms and conditions
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/cookies-and-privacy"
            >
              Cookies and privacy
            </a>
          </div>
        </div>
        <div
          class="mt-12 flex justify-between items-center flex-col gap-6 md:flex-row"
        >
          <div
            class="text-gray-500 flex gap-4 items-start"
          >
            <svg
              aria-label="Commerce Layer logomark"
              width="32"
            />
            <div
              class="text-xs leading-relaxed"
            >
              Copyright © 2022. Commerce Layer, Inc.
              <br />
              2965 Woodside Road, Woodside, CA 94062
              <br />
            </div>
          </div>
          <div
            class="text-gray-500 flex gap-4 items-start grayscale"
          >
            <svg />
            <svg />
            <svg />
            <svg />
            <svg />
            <svg />
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
